### DISCUSSION TITLE
Problems with recurring donations and periods using a custom payment gateway

### DISCUSSION THREAD

**Discussion by carazo** [impress-org/givewp#6780] [Developer Q&A]
I have created a gateway to enable RedSys payment (a Spanish one) into Give and all is working flawlessly except with recurring donations that are not based on a monthly subscription.

I use this hook: `add_action( 'give_daily_scheduled_events', array( $this, 'check_for_renewal_subscriptions' ), 1 );
`
Like the manual payments recurring donation gateway does and I do something like this to look for payments to be sent:
`function check_for_renewal_subscriptions() {
		global $wpdb;

		$subscriptions = $this->get_renewal_subscriptions();

		if ( ! empty( $subscriptions ) ) {
			foreach ( $subscriptions as $subscription ) {
				$add_renewal = false;

                // si la suscripción se paga un número determinado de veces, ver si se ha pagado menos veces de lo debido
				if ( $subscription->bill_times ) {
					$query    = $wpdb->prepare(
						"SELECT COUNT( id ) as count FROM $wpdb->posts WHERE post_status='give_subscription' AND post_type = 'give_payment' AND post_parent = %d",
						$subscription->parent_payment_id
					);
					$payments = $wpdb->get_results( $query );
					$payments = (int) $payments[0]->count;

					if ( $subscription->bill_times > ( $payments + 1 ) ) {
						$add_renewal = true;
					}
				} else { // si se paga de forma indefinida, seguro que hay que renovar
					$add_renewal = true;
				}

				if ( $add_renewal ) {
					$sub            = new Give_Subscription( absint( $subscription->id ) );
					$give_recurring = Give_Recurring();

					remove_action( 'give_recurring_add_subscription_payment', array( $give_recurring->emails, 'send_subscription_received_email' ), 10 );

                    $payment = $this->crear_pago_suscripcion( $sub );
                    $resultado_pago_suscripcion = $this->enviar_pago_suscripcion( $payment, $sub );

                    if( $resultado_pago_suscripcion == 'OK' ){
                        $this->confirmar_pago_suscripcion( $payment, $sub );
                        add_action( 'give_recurring_add_subscription_payment', array( $give_recurring->emails, 'send_subscription_received_email' ), 10, 3 );
                    }
                    else{
                        $payment->update_status( 'cancelled' );
                        $payment->add_note( 'Pago sucripción con RedSys falló por: ' . $resultado_pago_suscripcion );
                    }
				}
			}
		}
	}`

crear_pago_suscripcion creates a new payment and enviar_pago_suscripcion calls the web service of the bank.

I cannot understand why quarterly payments are made monthly.

**Comment by kjohnson** [comment]
@carazo most payment gateways do not support _quarterly_ subscriptions, so when a donor makes a quarterly donation GiveWP will convert that to the corresponding number of months.

For example:
- Every 1 Quarter becomes Every 3 Months
- Every 2 Quarters becomes Every 6 Months
- etc

**Comment by carazo** [comment]
No, this is not the reason. As you can see in the code I prepared for you, RedSys use a web service to send payments.

You can send subscriptions every hour, every minute, every day... as you wish.
